mava/components/tf/losses/sequence.py

```diff
@@ -27,45 +26,64 @@ def recurrent_n_step_critic_loss(
     q_values: tf.Tensor,
     target_q_values: tf.Tensor,
     rewards: tf.Tensor,
-    discounts: tf.Tensor,
     bootstrap_n: int,
-    loss_fn: Union[Type[trfl.td_learning], Type[losses.categorical]],
+    discount: float,
+    end_of_episode: tf.Tensor,
 ) -> tf.Tensor:
 
+    if type(q_values) == DiscreteValuedDistribution:
+        # Convert to tensor values
+        q_values = tf.reshape(q_values._mean(), rewards.shape)
+        target_q_values = tf.reshape(target_q_values._mean(), rewards.shape)
+
+    q_values = tf.squeeze(q_values)
+
+    target_q_values = tf.squeeze(target_q_values)
+    rewards = tf.squeeze(rewards)
+    check_rank([q_values, target_q_values, rewards], [2, 2, 2])
+
     seq_len = len(rewards[0])
-    assert bootstrap_n < seq_len
-    if type(q_values) != DiscreteValuedDistribution:
-        check_rank([q_values, target_q_values, rewards, discounts], [2, 2, 2, 2])
+
+    # d: discount * done
+    # bootstrap_n=1 is the normal return of Q_t-1 = R_t-1 + d * Q_t
+    # bootstrap_n=seq_len is the Q_t_1 = discounted sum of rewards return
+    assert 0 < bootstrap_n <= len(rewards[0])
 
     # Construct arguments to compute bootstrap target.
-    if type(q_values) != DiscreteValuedDistribution:
-        q_tm1 = q_values[:, 0:-bootstrap_n]
-        q_t = target_q_values[:, bootstrap_n:]
-
-        q_tm1, _ = combine_dim(q_tm1)
-        q_t, _ = combine_dim(q_t)
-    else:
-        q_tm1 = q_values
-        q_tm1.cut_dimension(axis=1, end=-bootstrap_n)
-
-        q_t = target_q_values
-        q_t.cut_dimension(axis=1, start=bootstrap_n)
-    n_step_rewards = rewards[:, :-bootstrap_n]
-    n_step_discount = discounts[:, :-bootstrap_n]
-    for i in range(1, bootstrap_n + 1):
-        n_step_rewards += rewards[:, i : i + seq_len - bootstrap_n]
+    q_tm1 = q_values
+
+    # The last values that rolled over do not matter because a
+    # mask is applied to it (hopefully).
+    q_t = tf.roll(target_q_values, shift=-bootstrap_n, axis=1)
 
+    # Create the end of episode mask.
+    ones_mask = tf.ones(shape=q_t[:, :-bootstrap_n].shape)
+    zeros_mask = tf.zeros(shape=q_t[:, -bootstrap_n:].shape)
+    eoe_mask = tf.concat([ones_mask, zeros_mask], axis=1)
+
+    q_tm1, _ = combine_dim(q_tm1)
+    n_step_rewards = rewards
+
+    # Pad the rewards so that rewards at the end can also be calculated.
+    zeros_mask = tf.zeros(shape=rewards[:, : bootstrap_n - 1].shape)
+    padded_rewards = tf.concat([rewards, zeros_mask], axis=1)
+    for i in range(1, bootstrap_n):
+        n_step_rewards += padded_rewards[:, i : i + seq_len] * math.pow(discount, i)
     n_step_rewards, _ = combine_dim(n_step_rewards)
-    n_step_discount, _ = combine_dim(n_step_discount)
 
-    critic_loss = loss_fn(
-        q_tm1,
-        n_step_rewards,
-        n_step_discount,
-        q_t,
+    q_t, _ = combine_dim(q_t)
+    eoe_mask, _ = combine_dim(eoe_mask)
```

Comment:
I know acme/trfl have these short cryptic names, but I'm wondering if we shouldn't perhaps opt for fuller names that are more descriptive. It took one or two seconds for me to get that `eoe` probably means `end-of-episode`. I think being more descriptive can improve code readability. What do you think @DriesSmit and @KaleabTessera ?

Reply:
Agreed. I fixed it :)